docs/design/src/lib/shared_sem.c

```diff
@@ -0,0 +1,35 @@
+#include "docs/design/include/mcmini/shared_sem.h"
+#include "docs/design/include/mcmini/wrappers.h"
+
+void shared_sem_init(shared_sem_ref ref) {
+  if (!ref) return;
+  __real_sem_init(&ref->runner_sem, 1, 0);
+    __real_sem_init(&ref->pthread_sem, 1, 0);
+}
+
+void shared_sem_destroy(shared_sem_ref ref) {
+  if (!ref) return;
+  __real_sem_destroy(&ref->runner_sem);
+  __real_sem_destroy(&ref->pthread_sem);
+}
+
+void shared_sem_post_for_thread(shared_sem_ref ref) {
+  if (!ref) return;
+  __real_sem_post(&ref->pthread_sem);
+}
+
+void shared_sem_wait_for_thread(shared_sem_ref ref) {
+  if (!ref) return;
+  __real_sem_wait(&ref->pthread_sem);
+}
+
+void shared_sem_post_for_runner(shared_sem_ref ref) {
+  if (!ref) return;
+  __real_sem_post(&ref->runner_sem);
+}
+
+void shared_sem_wait_for_runner(shared_sem_ref ref) {
+  if (!ref) return;
+  __real_sem_wait(&ref->runner_sem);
+}
```

Comment:
It's a bit confusing to use both `thread` and `runner` since `runner` really refers to a thread (or a process etc.). One of these corresponds to the _scheduler_ and the other the _thread_ waiting to later be scheduled.

Reply:
It's probably good to define `awaits_scheduler` e.g. that performs the post and then the wait at the same time so we don't only call one or the other (we need `post + wait` or `wait + post` on the other side)